### PR TITLE
Updated Weights & Biases logging + Made codebase pip installable

### PR DESCRIPTION
**Pull request by soumik12345**
Changes included in this PR:
- [x] Added use of the [`WandbMetricsLogger`](https://docs.wandb.ai/guides/integrations/keras#experiment-tracking-with-wandbmetricslogger) callback logging training metrics to Weights & Biases.
- [x] Added use of [`WandbModelCheckpoint`](https://docs.wandb.ai/guides/integrations/keras#model-checkpointing-using-wandbmodelcheckpoint) for logging model checkpoints as [Weights & Biases artifacts](https://docs.wandb.ai/guides/artifacts).
- [x] Added an additional callback `QualitativeValidationCallback` for visualizing the generations corresponding to multiple prompts in an epoch-wise manner using [Weights & Biases Table](https://docs.wandb.ai/guides/data-vis).
- [x] Updated the `validation_prompt` argument in the training script to accept multiple prompts for visualizing the generations during training.
- [x] Made the codebase pip installable

Todos:
- [x] Make sure to log the results corresponding to all prompts in both table and media panel.
- [x] Update README to reflect all the changes.

**Sample Weights & Biases run:** https://wandb.ai/geekyrakshit/dreambooth-keras/runs/h0kdq2dw

### CODE REVIEW

**Comment by sayakpaul** [comment]
@soumik12345 thanks for your PR!

I think we would want to keep the codebase a bit more lightweight for people to be able to navigate around easily and customize as per their needs. 

So, with that in mind, I would request to introduce the changes _only_ pertaining to WandB logging (images). Also, please keep the code-lints unchanged.

**Comment by soumik12345** [comment]
Hi @sayakpaul 
I updated the PR as per your request. Here's a sample wandb run: https://wandb.ai/geekyrakshit/dreambooth-keras/runs/4wydmf8w

**Comment by sayakpaul** [comment]
> Hi @sayakpaul I updated the PR as per your request. Here's a sample wandb run: https://wandb.ai/geekyrakshit/dreambooth-keras/runs/4wydmf8w
> 
> I will edit the readme to include the changes, once you approve the changes.

I still don't see the prompt in the validation panel.

**Comment by soumik12345** [comment]
Hi @sayakpaul, I updated the logging of generated images from validation prompt on W&B panel. Here's a sample run: https://wandb.ai/geekyrakshit/dreambooth-keras/runs/8shvurfp

File: src/utils.py [hunk 1]
```diff
@@ -1,33 +1,92 @@
...
+    for prompt in tqdm(prompts):
+        images_dreamboothed = sd_model.text_to_image(prompt, batch_size=num_imgs_to_gen)
+        wandb.log(
+            {
+                f"validation/Prompt: {prompt}": [
+                    wandb.Image(PIL.Image.fromarray(image), caption=f"{i}: {prompt}")
+                    for i, image in enumerate(images_dreamboothed)
+                ]
+            }
+        )
```
Comment (sayakpaul) [inline src/utils.py:34]: Nice!

File: train_dreambooth.py [hunk 2]
```diff
@@ -148,32 +166,44 @@ def run(args):
...
+                ckpt_path_prefix, save_weights_only=True, monitor="loss", mode="min"
+            ),
+            QualitativeValidationCallback(
+                img_heigth=args.img_resolution,
+                img_width=args.img_resolution,
+                prompts=validation_prompts,
+                num_imgs_to_gen=args.num_images_to_generate,
+            ),
+        ]
+        if args.log_wandb
```
Comment (sayakpaul) [inline train_dreambooth.py:187]: Please refactor it such that the else block becomes more pronounced. No comprehensions here, please.

File: train_dreambooth.py [hunk 3]
```diff
@@ -148,32 +166,44 @@ def run(args):
...
+        else [
+            tf.keras.callbacks.ModelCheckpoint(
+                ckpt_path_prefix, save_weights_only=True, monitor="loss", mode="min"
             )
+        ]
+    )
```
Comment (sayakpaul) [inline train_dreambooth.py:193]: The problem with only saving the low-loss checkpoints in DreamBooth is that the loss curves are not that informative if you take a look at them (think of vanilla GAN training loss curves).

File: train_dreambooth.py [hunk 4]
```diff
@@ -148,32 +166,44 @@ def run(args):
...
-        wandb.init(project="dreambooth-keras", name=run_name, config=vars(args))
-        if args.validation_prompt is None:
-            args.validation_prompt = (
-                f"A photo of {args.unique_id} {args.class_category} in a bucket"
+    callbacks = (
+        [
+            WandbMetricsLogger(log_freq="batch"),
+            DreamBoothCheckpointCallback(
+                ckpt_path_prefix, save_weights_only=True, monitor="loss", mode="min"
+            ),
```
Comment (sayakpaul) [inline train_dreambooth.py:179]: Same as https://github.com/sayakpaul/dreambooth-keras/pull/22/files#r1105388125

File: train_dreambooth.py (hunk 3 above)
Comment (sayakpaul) [reply train_dreambooth.py:193]: So, this is why I just logged weights after the end of training. 

In my experience, the low-loss checkpoints often lead to inferior results.

**Review by sayakpaul — COMMENTED** [review]
Thanks for the changes.

File: train_dreambooth.py (hunk 3 above)
Comment (soumik12345) [reply train_dreambooth.py:193]: I'll remove the monitor in that case.

File: train_dreambooth.py (hunk 2 above)
Comment (sayakpaul) [reply train_dreambooth.py:185]: Not addressed.

**Review by sayakpaul — COMMENTED** [review]
One change remains.

File: train_dreambooth.py (hunk 2 above)
Comment (soumik12345) [reply train_dreambooth.py:185]: On it

Comment (soumik12345) [reply train_dreambooth.py]: @sayakpaul made the change.

File: train_dreambooth.py [hunk 5]
```diff
@@ -171,22 +171,28 @@ def run(args):
...
+            DreamBoothCheckpointCallback(ckpt_path_prefix, save_weights_only=True)
+        )
+        # perform inference on validation prompts at the end of every epoch and
+        # log the resuts to a Weights & Biases table
+        callbacks.append(
             QualitativeValidationCallback(
                 img_heigth=args.img_resolution,
                 img_width=args.img_resolution,
                 prompts=validation_prompts,
                 num_imgs_to_gen=args.num_images_to_generate,
```
Comment (sayakpaul) [inline train_dreambooth.py:189]: Perfect 💯

**Review by sayakpaul — APPROVED** [review]

File: train_dreambooth.py [hunk 6]
```diff
@@ -148,32 +166,46 @@ def run(args):
...
             )
+        )
+    else:
+        callbacks.append(
+            tf.keras.callbacks.ModelCheckpoint(ckpt_path_prefix, save_weights_only=True)
+        )
 
+    train(dreambooth_trainer, train_dataset, args.max_train_steps, callbacks)
+
+    if args.log_wandb:
```
Comment (deep-diver) [inline train_dreambooth.py:199]: nit: maybe it is better to reuse `StableDiffusion` from `QualitativeValidationCallback` since it is always there under `args.log_wandb=True` condition.

**Review by deep-diver — CHANGES_REQUESTED** [review]

**Review by deep-diver — APPROVED** [review]

**Comment by sayakpaul** [comment]
Feel free to add a note of these changes to README, @soumik12345! Thank you.

**Comment by soumik12345** [comment]
> Feel free to add a note of these changes to README, @soumik12345! Thank you.

@sayakpaul updated the README to reflect the changes.

**Review by sayakpaul — APPROVED** [review]
